lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,230 @@
+# frozen_string_literal: true
+require "json"
+require "reverse_markdown"
+
+module Licensed
+  module Sources
+    # Only supports ProjectReference (project.assets.json) style restore used in .NET Core.
+    # Does not currently support packages.config style restore.
+    class NuGet < Source
+      def self.type
+        "nuget"
+      end
+
+      class NuGetDependency < Licensed::Dependency
+        LICENSE_FILE_REGEX = /<license\s*type\s*=\s*\"\s*file\s*\"\s*>\s*(.*)\s*<\/license>/ix.freeze
+        LICENSE_URL_REGEX = /<licenseUrl>\s*(.*)\s*<\/licenseUrl>/ix.freeze
+        PROJECT_URL_REGEX = /<projectUrl>\s*(.*)\s*<\/projectUrl>/ix.freeze
+
+        def initialize(name:, version:, path:, search_root: nil, metadata: {}, errors: [])
+          super(name: name, version: version, path: path, search_root: search_root, metadata: metadata, errors: errors)
+          @metadata["homepage"] = project_url if project_url
+        end
+
+        def nuspec_path
+          name = @metadata["name"]
+          File.join(self.path, "#{name}.nuspec")
+        end
+
+        def nuspec_contents
+          return unless nuspec_path
+          @nuspec_contents ||= File.read(nuspec_path)
+        end
+
+        def project_url
+          return unless nuspec_contents
+          @project_url ||= begin
+            match = nuspec_contents.match PROJECT_URL_REGEX
+            match[1] if match && match[1]
+          end
+        end
+
+        def project_files
```

Comment:
it would duplicate a bit of logic, but if you find a recognized license in a `Licensee::ProjectFiles::PackageManagerFile` and its contents match `Licensee::Matchers::Nuget::LICENSE_URL_REGEX` then you could add a license file like
```
# strip copyright clauses and any extra newlines
text = file.license.text
                   .lines
                   .reject { |l| l =~ Licensee::Matchers::Copyright::REGEX }
                   .join
                   .gsub(/\n\n\n/, "\n\n")

Licensee::ProjectFiles::LicenseFile.new(text, { uri: <matched url> })
```

The text generation was copied from [dependency.rb](https://github.com/github/licensed/blob/b5e8cb9b7257d7102a656cb6e242ab14b3c3287c/lib/licensed/dependency.rb#L140-L158).

Reply:
I assume you mean checking that the contents of the `PackageManagerFile` itself match against `LICENSE_URL_REGEX`? 

If so, it's possible that the `PackageManagerFile` license actually came from `Licensee::Matchers::Nuget::LICENSE_REGEX` (`license` expression) and is potentially a different license than `licenseUrl`, so I think you'd also want to check that the package file contents *doesn't* match against `LICENSE_REGEX`.

That would definitely be better though than the current auto-generated source. I was wondering if this sort of change might belong in licensee, but I think it would require some non-trivial refactoring of `PackageManagerFile` which is why I didn't.

If you're ok without this change for now, it might make sense to do this as a separate PR (which I'd be happy to do).

Reply:
> I assume you mean checking that the contents of the PackageManagerFile itself match against LICENSE_URL_REGEX?

Yep 👍 

> If so, it's possible that the PackageManagerFile license actually came from Licensee::Matchers::Nuget::LICENSE_REGEX (license expression) and is potentially a different license than licenseUrl,

Oh, does the licensee change not look for both of them and report other if they don't match?

> If you're ok without this change for now, it might make sense to do this as a separate PR (which I'd be happy to do).

yeah for sure its fine not to block this one.  the impact being that you might get a bunch of minimal changes to your cached data if the followup PR is put into a future release

Reply:
> Oh, does the licensee change not look for both of them and report other if they don't match?

No. Perhaps it should have, but I think I saw that others don't do that (e.g. gemspec matcher doesn't ensure `license` and `licenses` don't conflict), and `licenseUrl` is [officially deprecated](https://docs.microsoft.com/en-us/nuget/reference/nuspec#licenseurl) so it seemed reasonable to ignore `licenseUrl` if `license` is set.

And in the case where `licenseUrl` is also set but not a recognized URL, you definitely wouldn't want to report `other`.

Reply:
Per discussion, will leave as is and file an issue to consider for the future.